cmd/juju/application/deployer/bundlehandler.go

```diff
@@ -653,38 +653,39 @@ func (h *bundleHandler) addCharm(change *bundlechanges.AddCharmChange) error {
 	if err != nil {
 		return errors.Annotatef(err, "cannot resolve %q", ch.Name)
 	}
-	switch {
-	case url.Series == "bundle" || resolvedOrigin.Type == "bundle":
+
+	if url.Series == "bundle" || resolvedOrigin.Type == "bundle" {
 		return errors.Errorf("expected charm, got bundle %q %v", ch.Name, resolvedOrigin)
-	case resolvedOrigin.Base.Channel.Empty():
-		modelCfg, workloadSeries, err := seriesSelectorRequirements(h.deployAPI, h.clock, url)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		selector := seriesSelector{
-			charmURLSeries:      url.Series,
-			seriesFlag:          change.Params.Series,
-			supportedSeries:     supportedSeries,
-			supportedJujuSeries: workloadSeries,
-			conf:                modelCfg,
-			fromBundle:          true,
-		}
+	}
 
-		// Get the series to use.
-		chSeries, err := selector.charmSeries()
-		if err != nil {
-			return errors.Trace(err)
-		}
-		url = url.WithSeries(chSeries)
+	modelCfg, workloadSeries, err := seriesSelectorRequirements(h.deployAPI, h.clock, url)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-		// TODO(juju3) - use os/channel, not series
-		base, err := series.GetBaseFromSeries(chSeries)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		resolvedOrigin.Base = base
-		logger.Tracef("Using channel %s from %v to deploy %v", resolvedOrigin.Base.String(), supportedSeries, url)
+	selector := seriesSelector{
+		charmURLSeries:      url.Series,
+		seriesFlag:          change.Params.Series,
+		supportedSeries:     supportedSeries,
+		supportedJujuSeries: workloadSeries,
+		conf:                modelCfg,
+		fromBundle:          true,
+	}
+
+	// Get the series to use.
+	chSeries, err = selector.charmSeries()
+	if err != nil {
+		return errors.Annotatef(err, "failed to upload charm %s", chParams.Charm)
```

Comment:
s/chParams.Charm/ch.Name/

Use the charm name, rather than URL.

Reply:
Agreed